pygen/pygen_src/isa/riscv_instr.py

```diff
@@ -0,0 +1,341 @@
+            self.include_reg = cfg.invalid_priv_mode_csrs
+        else:
+            if(cfg.init_privileged_mode == "MACHINE_MODE"):    # Machine Mode
+                self.include_reg.append("MSCRATCH")
+            elif(cfg.init_privileged_mode=="SUPERVISOR_MODE"): # Supervisor Mode
+                self.include_reg.append("SSCRATCH")
+            else:                                              # User Mode
+                self.include_reg.append("USCRATCH") 
+
+    def get_rand_instr(self):
+        pass #TODO
+
+    def get_load_store_instr(self, load_store_instr):
+        instr_h = riscv_instr()
+        if(len(load_store_instr) == 0):
+            load_store_instr = self.instr_category["LOAD"] + self.instr_category["STORE"]
+        self.idx = random.randrange(0,len(load_store_instr)-1)    
+        name = load_store_instr[self.idx]
+        instr_h = copy(self.instr_template[name]) #Shallow copy for all relevant fields to improve performance
+        return instr_h
+
+    def get_instr(self, name):
+        if (not self.instr_template.get(name)):             
+            logging.critical("Cannot get instr %s",name)
+        instr_h = copy(self.instr_template[name])
+        return instr_h
+
+    def set_rand_mode(self):
+        # rand_mode setting for Instruction Format
+
+        if(self.format.name == "R_FORMAT"):
+            self.has_imm = 0 
+        if(self.format.name=="I_FORMAT"):
+            self.has_rs2 = 0
+        if(self.format.name in ["S_FORMAT", "B_FORMAT"]):
+            self.has_rd = 0
+        if(self.format.name in ["U_FORMAT", "J_FORMAT"]):
+            self.has_rs1 = 0
+            self.has_rs2 = 0
+
+        # rand_mode setting for Instruction Category
+        if(self.category.name == "CSR"):
+            self.has_rs2 = 0
+            if(self.format.name == "I_FORMAT"):
+                self.has_rs1 = 0
+
+    def pre_randomize(self):
+        pass #TODO
+    
+    def set_imm_len(self):
+        if(self.format.name in  ["U_FORMAT","J_FORMAT"]):
+            self.imm_len = 20
+        elif(self.format.name in ["I_FORMAT", "S_FORMAT", "B_FORMAT"]):
+            if(self.imm_type.name == "UIMM"):
+                self.imm_len = 5
+            else:
+                self.imm_len = 11
+       	self.imm_mask = self.imm_mask << self.imm_len
+
+   
+    def extend_imm(self):
+        sign = 0
+        self.imm = self.imm << (32 - self.imm_len)
+        sign = self.imm.bin[0:1:1] # sign = imm[31]
+        self.imm = self.imm >> (32 - self.imm_len)
+        #Signed extension
+        if((sign and not(self.format == "U_FORMAT")) or (self.imm_type in ["UIMM", "NZUIMM"])):
+        	self.imm = self.imm_mask | self.imm
+
+
+    def post_randomize(self):
+        self.extend_imm()
+        self.update_imm_str()
+        
+    def convert2asm(self):
+        pass #TODO
+
+    def get_opcode(self):
+    	count =0
+    	if(self.instr_name.name == "LUI"):
+    		count +=1
+    		return (BitArray(uint=55 , length = 7).bin)
```

Comment:
It is better to use BitArray with `hex` argument

Reply:
Can we take BitArray with `bin` arguments instead of `hex`?
As `get_opcode()` function  returns the 7 bit binary value.

Reply:
A side note, we can port over get_opcode much later, this is only needed for illegal/hint instruction test.